Neutrino component

preshanth opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request introduces the neutrino component trait class into cosmology

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes a portion of #17150

CC  @nstarman
<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -69,15 +84,9 @@
-KOMATSU_K: Final = 0.3173
+# Note: Neutrino physics constants (NEUTRINO_FERMI_DIRAC_CORRECTION, KOMATSU_P, etc.)
+# are now defined in astropy.cosmology._src.traits.neutrino and imported above.
+# They are re-exported from this module for backward compatibility.
```
same.

preshanth commented on astropy/cosmology/_src/traits/neutrino.py:
```diff
@@ -0,0 +1,317 @@
+        """
+        # The constant in front is (4/11)^1/3 -- see any cosmology book for an
+        # explanation -- for example, Weinberg 'Cosmology' p 154 eq (3.1.21).
+        return 0.7137658555036082 * self.Tcmb0
```
Done.

preshanth commented:
@nstarman If the total component gets merged. I will work on the docs and changes and changes.

preshanth commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -1,7 +1,4 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
-
-__all__ = ["FLRW", "FlatFLRWMixin"]
```
Nope its not in my working branch. Will see where it came from.
```python
# Licensed under a 3-clause BSD style license - see LICENSE.rst
import inspect
import warnings
from dataclasses import field
from functools import cached_property
from inspect import signature
from math import floor, pi, sqrt
from numbers import Number
from typing import Any, Final, NamedTuple, TypeVar, overload
```

nstarman commented on astropy/cosmology/_src/traits/neutrino.py:
```diff
@@ -0,0 +1,318 @@
+        # you need to update there too.
+
+        # The massive and massless contribution must be handled separately
+        # But check for common cases first
```
```suggestion
        # No massive neutrinos
```

nstarman commented on astropy/cosmology/_src/traits/neutrino.py:
```diff
@@ -0,0 +1,318 @@
+        if not self._nu_info.has_massive_nu:
+            return NEUTRINO_FERMI_DIRAC_CORRECTION * self.Neff * np.ones_like(z)
+
+        curr_nu_y = self._nu_info.nu_y / (1.0 + np.expand_dims(z, axis=-1))
```
```suggestion
        # At least some massive neutrinos
        curr_nu_y = self._nu_info.nu_y / (1.0 + np.expand_dims(z, axis=-1))
```

nstarman reviewed (commented):
A few comments. Overall, looking good.

nstarman commented:
@preshanth can you rebase and change the whatsnew to 8.0.rst ?

preshanth commented:
@nstarman I could use a suggestion of which way to go.  The `_nu_info` method I think should be an abstract method in the trait. Then I want to move the `_nu_info` into a method inside flrw and then I can show a real custom neutrino cosmology. Currently that is not that case. Do you think that makes sense to you. If you say yes I will go ahead and implement that. 
I have rebased and fixed the `whatsnew` issue. Just was realizing that the docs examples don't really show an actual neutrino component trait.

preshanth commented:
@nstarman I rebased and did a bit more than what I was planning. Cause it was annoying me. Have a look at `neutrino.py` I introduced another private class there made the neutrinos more trait like and independent. I also showed an FLRW like class in the docs as an example.

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -1,19 +1,16 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
-
-__all__ = ["FLRW", "FlatFLRWMixin"]
```
Please restore this `__all__`.

`__all__ = ["FLRW", "FlatFLRWMixin"]`

nstarman commented on docs/whatsnew/8.0.rst:
```diff
@@ -65,6 +65,136 @@ After:
+:class:`~astropy.cosmology.traits.NeutrinoComponent`,
+:class:`~astropy.cosmology.traits.MatterComponent`, and
+:class:`~astropy.cosmology.traits.CriticalDensity` traits provide the scale factor, the temperature or the CMB, the Dark Energy component, and the Dark Matter component,
+respectively.
```
```suggestion
behaviors. For example, :class:`~astropy.cosmology.traits.NeutrinoComponent` provides the neutrino energy density.
```

nstarman reviewed (commented):
Thanks @preshanth. This is definitely the most challenging of the traits.
It looks like a fair bit of stuff got transferred in the rebase that needs some cleanup.

I actually think it may be best to start simpler with this one. I'm not sure if we want to have `_nu_info` on all the classes, and not just as an implementation detail of `FLRW`. Like in the other trait classes, I suggest starting this PR with abstract methods in the trait that get overridden in the FLRW subclass. Then we can iterate to make there be fewer abstract methods. 
Sorry, I know that's a bit of a different tack from the current direction of this PR.

nstarman commented on docs/whatsnew/8.0.rst:
```diff
@@ -65,6 +65,136 @@ After:
+...     TotalComponent,
+...     NeutrinoComponent,
+... )
+>>> class CustomStandardCosmology(Cosmology, HubbleParameter, ScaleFactor, TemperatureCMB,
```
If that's easier, sure! Happy to review either way.

preshanth replied:
Let me try if I fail this time around. I will start a new PR clean. This is a fun one. You can tell I was playing around a bit too  much.

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -70,28 +69,29 @@
-# what this does. However, this is modified to handle multiple neutrino masses by
-# computing the above for each mass, then summing
+# Neutrino physics constants for FLRW calculations
+# See Komatsu et al. 2011, eq 26 and the surrounding discussion
```
Please revert.

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -70,28 +69,29 @@
 NEUTRINO_FERMI_DIRAC_CORRECTION: Final = 0.22710731766  # 7/8 (4/11)^4/3
-
-# These are purely fitting constants -- see the Komatsu paper
+# Komatsu fitting constants
```
Please revert.

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -70,28 +69,29 @@
 
+##############################################################################
+# NeutrinoInfo - FLRW-specific implementation detail
 ##############################################################################
```
```suggestion
##############################################################################
# NeutrinoInfo - FLRW-specific implementation detail
```

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -85,13 +85,17 @@
+
+# NeutrinoInfo - FLRW-specific implementation detail
+
 ##############################################################################
```
```suggestion
##############################################################################
# NeutrinoInfo - FLRW-specific implementation detail
```

preshanth commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -507,109 +526,7 @@ def Otot(self, z: u.Quantity | ArrayLike) -> FArray:
-        If they have different masses, a similar term is calculated for each
-        one. Note that ``f`` has the asymptotic behavior :math:`f(0) = 1`. This
-        method returns :math:`0.2271 f` using an analytical fitting formula
-        given in Komatsu et al. 2011, ApJS 192, 18.
```
Do you want me to put in an example like in your other PR ?

preshanth replied:
Okay I have put that in. Making sure the docs pass and will push up.

nstarman replied:
Are you using a raw string? `r"""` ?

nstarman commented:
Sorry. Recently merged PRs mean you rebase to rebase this one.

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -1,6 +1,6 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
-__all__ = ("FLRW", "FlatFLRWMixin")
+__all__ = ["FLRW", "FlatFLRWMixin"]
```
```suggestion
__all__ = ("FLRW", "FlatFLRWMixin")
```
